src/platform/forms-system/src/js/fields/FileField.jsx

```diff
@@ -29,25 +40,119 @@ export default class FileField extends React.Component {
     }
   }
 
-  onAddFile = (event, index = null) => {
+  focusAddAnotherButton = () => {
+    // focus on span pseudo-button, not the label
+    focusElement(`#${this.props.idSchema.$id}_add_label span`);
+  };
+
+  /**
+   * @typedef Files~state - File state object
+   * @type {object}
+   * @property {object|null} file - upload API supplied File object: see
+   *   https://developer.mozilla.org/en-US/docs/Web/API/File
+   * @property {number} index - index matching formData array position
+   * @property {boolean} isEncrypted - detected encrypted state of a PDF file
+   * @property {boolean} isTemporary - indicates state data that needs needs a
+   *   placeholder
+   */
+  /**
+   * Update file state
+   * @param {object|null} file - API supplied File object; store the entire File
+   *   object because it needs to be passed to the `uploadRequest` function
+   * @param {number} index - index of file in formData
+   * @param {boolean} isEncrypted - encrypted state
+   */
+  setFileState = ({ file, index, isEncrypted }) => {
+    // remove already uploaded files from state
+    const files = this.state.files.filter(
+      fileObj =>
+        this.props.formData.findIndex(
+          fileData => fileData.name === fileObj.file.name,
+        ) < 0,
+    );
+
+    const existingIndex = files.findIndex(fileData => fileData.index === index);
+    const data = { file, index, isEncrypted, isTemporary: true };
+    if (existingIndex > -1) {
+      files[existingIndex] = data;
+      this.setState({ files });
+    } else {
+      this.setState({ files: [...files, data] });
+    }
+  };
+
+  onSubmitPassword = (file, index, password) => {
+    if (file && password) {
+      this.onAddFile({ target: { files: [file] } }, index, password);
+    }
+  };
+
+  processFile = async (file, index) => {
+    const { uiSchema } = this.props;
+    return checkForEncryptedPdf(file, uiSchema)
+      .then(isEncrypted => {
+        if (isEncrypted) {
+          this.setFileState({ file, index, isEncrypted: true });
+        }
+        return isEncrypted;
+      })
+      .catch(() => {
+        // This _should_ only be called if a file is deleted after the
+        // user selects it for upload
+        this.setFileState({ index, isEncrypted: false });
+        return false;
+      });
+  };
```

Comment:
Is `file` guaranteed to not be null here? The JSDoc for `setFileState` says that _that_ `file` arg can be null, and with the removal of the try/catch in `checkForEncryptedPdf`, I think a null `file` will throw an error. Also, when `onAddFile` is called from `onSubmitPassword` it looks like `file` has the possibility to be null.

So if `checkForEncryptedPdf` throws an error, we may end up in the catch block here, which is only supposed to happen if a file is deleted. Are those the same scenario?

I'm also not sold on the `processFile` name. Would `setEncryptedState` be more descriptive?

Reply:
- The `file` could have been `null` in the `state` data, but I ended up removing all the `state` data and added the pre-uploaded file directly into the `formData`.

The `onAddFile` is wrapped in a function that checks for the `file`, and I've added an additional check that the file includes a `name` that ends in `pdf`. Then `processFile` is called. All of this happens immediately after the user selects a file, so the odds of `file` being `null` at this point would be essentially non-existent. The problem is that `file` may throw an error _after_ the user enters a password and clicks the "Add password" button if the file was deleted in that period of time - in which case `file` _is_ defined, but does not reference an actual file. I just tried this scenario and got a Network error

![Screen Shot 2020-10-28 at 12 42 19 PM](https://user-images.githubusercontent.com/136959/97475554-65f07180-191b-11eb-8d4b-a7f09bba7774.png)

Lastly, I'll rename this function to make it more clear